### DISCUSSION TITLE
Understanding the pandoc AST "schema" better

### DISCUSSION THREAD

**Discussion by alexanderjulmer** [jgm/pandoc#9436] [Q&A] [answered]
I am trying to understand which elements pandoc allows to be nested and which it will un-nest when converting. 

I want to understand this because I am converting a pandoc AST document to an even more restrictive document schema. For this, I would like to understand where I could add mitigations. For example, the target format does not allow inline images, but those could be inserted as siblings behind the containing block.

I found [`Text.Pandoc.Definition`](https://hackage.haskell.org/package/pandoc-types-1.23.1/docs/Text-Pandoc-Definition.html#t:Block) where I read the 'Block' section to explain what is allowed in which block, e.g. `BulletList [[Block]]` would allow _any_ block inside a bullet list. Yet, when given a heading in a bullet list (I don't know why you would want that, but Word allows it), pandoc will return:

````
[
  BulletList [ [ Para [ Str "Asdasdasd" ] ] ]
  , Header 1 ( "asdasdasd" , [] , [] ) [ Str "Asdasdasd" ]
  , BulletList
      [ [ Para [ Str "Asdasdasda" ] ]
      , [ Para [ Str "asdasdasd" ] ]
      ]
]
````

This leads me to believe that the underlying schema is actually stricter but I cannot find any more information. Is there a resource I could look at?

Your help is greatly appreciated.

**Comment by jgm** [comment, marked as answer]
Yes, a heading is allowed in a bullet list in the pandoc AST.
The docx reader may be set up not to return this, but :
```
 % pandoc -f markdown -t native
- # Hi
- ok
[ BulletList
    [ [ Header 1 ( "hi" , [] , [] ) [ Str "Hi" ] ]
    , [ Plain [ Str "ok" ] ]
    ]
]
```

**Reply by alexanderjulmer** [reply]
Thank you for clarifying this. Then I'll use the schema in [Text.Pandoc.Definition](https://hackage.haskell.org/package/pandoc-types-1.23.1/docs/Text-Pandoc-Definition.html#t:Block) as the basis for my implementation.

Is there documentation of how the docx reader is set up? It also has that interesting behaviour of converting lists to blockquotes when the bullet is slightly indented. I would like to understand that better as it often appears in contracts and other legal documents.

And since you're here already: Thank you for your wonderful work. We couldn't do what we do without pandoc!

**Reply by jgm** [reply]
> Is there documentation of how the docx reader is set up? 

Sorry, no.  You'd have to dig into the source.
